### ISSUE TITLE
GitHubUpdateManager() should use ConfigureAwait(false)

### ISSUE THREAD

**Issue by RyusukeFujita** [clowd/Clowd.Squirrel#18]
This is just a suggestion.

[GitHubUpdateManager](https://github.com/clowd/Clowd.Squirrel/blob/master/src/Squirrel/UpdateManager.Factory.cs) uses a few "await", but it should be using ConfigureAwait(false):

Suggested Code Change:

```C#
using (var client = new HttpClient() { BaseAddress = baseAddress }) {
  client.DefaultRequestHeaders.UserAgent.Add(userAgent);
  var response = await client.GetAsync(releasesApiBuilder.ToString()).ConfigureAwait(false);
  response.EnsureSuccessStatusCode();

  var releases = SimpleJson.DeserializeObject<List<Release>>(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
````

This change will avoid a deadlock in a scenario where the caller of GitHubUpdateManager() waits synchronously such as:

```C#
  var mgr = GetUpdateManagerAsync().GetAwaiter().GetResult();
```

If the caller uses the UI thread to invoke like this, the UI thread will wait for the task to finish, while GitHubUpdateManager() awaits a task and resume on the same UI thread.  However, the UI thread cannot resume GitHubUpdateManager() because it's already waiting.

**Comment by caesay** [comment]
Hi, thanks for the suggestion. Presumably the entire library suffers from this problem, as it's quite heavy with async/await. Is this a common issue, and what is your suggestion for the rest of the code?

**Comment by RyusukeFujita** [comment]
> what is your suggestion for the rest of the code?

In my opinion, generally speaking the library code should use `.ConfigureAwait(false)` whenever possible (i.e., when there's no need to await & resume with the same thread) since the library has no way of knowing how it's going to be used (the caller might invoke async function synchronously with UI thread as described).

I think this kind of issue is common. For example, ...

```cs
// Suppose this is the library user and this code gets invoked with UI thread.
private void button1_Click(object sender, EventArgs e)
{
    string s = DoSomeAsync().Result;
}

// Suppose this is a library code
private async Task<string> DoSomeAsync()
{
    await Task.Delay(1000); 
    return "test";
}
```

This causes a deadlock with the exact same reason: When UI thread hits `await Task.Delay(1000)`, it returns to `button1_Click()` and it will wait until this task completes.  After `Task.Delay()` completes, CLR will schedule to resume `DoSomeAsync()` with UI thread. However, UI thread is already waiting and thus it never resumes.  

This might be a silly example, but if the library code used `await Task.Delay(1000).ConfigureAwait(false)`, deadlock wouldn't occur.

By the way, even if you don't fix the code, the caller can still workaround the issue by manually offloading to a worker thread:

e.g. 
```cs
// Suppose this code is executed by UI thread.
var mgr = Task.Run(async () => await UpdateManager.GitHubUpdateManager(Config.UpdateUri)).GetAwaiter().GetResult();
```

With this, `GitHubUpdateManager()` runs in a background thread (with no synchronization context). It can await & resumes without any thread affinity.

So, workaround is simple.  Thus, this post is just a "suggestion".  But I thought it'd be nice that the caller can simply do like this without deadlocking:

```cs
var mgr = GetUpdateManagerAsync().GetAwaiter().GetResult();
```

**Comment by caesay** [comment]
In your example code, 

```cs
// Suppose this is the library user and this code gets invoked with UI thread.
private void button1_Click(object sender, EventArgs e)
{
    string s = DoSomeAsync().Result;
}
```

for what reason would you want to use `.Result` instead of marking the function async and using await? presumably calling `.Result` will cause the UI to freeze until the task is complete.

**Comment by RyusukeFujita** [comment]
Thanks, Caelan.
In this example, there's no good reason to do that (that's why I called it a "silly example").

The point is that library has no way of knowing how it's going to be used. Therefore, I think it's better to use `.ConfigureAwait(false)` whenever possible (Just glancing at [GitHubUpdateManager](https://github.com/clowd/Clowd.Squirrel/blob/master/src/Squirrel/UpdateManager.Factory.cs), I didn't see any reason not to use it).

**Comment by chm-tm** [comment]
From [what I read](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2007), using `.ConfigureAwait` in libraries is strongly recommended.

While this clutter annoys me in my library code, avoiding dead locks is definitively worth it. And I suffered from quite some of them due to missed `.ConfigureAwait(false)`s. So, praise for [CA2007](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2007)!

**Comment by caesay** [comment]
Thanks for the tip about CA2007, very helpful. This has been addressed in 0b8bcc19d650da1592a8cc26f7861b55a5a179bd and will be in the next release.

**Comment by caesay** [comment]
This is now released [in 2.7.34-pre](https://github.com/clowd/Clowd.Squirrel/releases/tag/2.7.34-pre)

**Comment by RyusukeFujita** [comment]
Thanks Caelan!
